File: vol1/ch3/LeetCode20.cpp
class Solution {
public:
    bool isValid(string s) {
        int n=s.size();
        if(n==0) return true;
        stack<char> st;
        for(int i=0;i<n;i++) {
            if(s[i]=='(' || s[i]=='[' || s[i]=='{')
                st.push(s[i]);
            else if(s[i]==')') {
                if(st.empty()) return false;
                if(st.top()!='(') return false;
                st.pop();
            }
            else if(s[i]==']') {
                if(st.empty()) return false;
                if(st.top()!='[') return false;
                st.pop();
            }
            else {                       //s[i]='}'
                if(st.empty()) return false;
                if(st.top()!='{') return false;
                st.pop();
            }
        }
        return st.empty();
    }
};

File: vol1/ch3/LeetCode20.py
class Solution:
    def isValid(self, s: str) -> bool:
        n=len(s)
        if n==0:return True
        st=[]
        for i in range(0,n):
            if s[i]=='(' or s[i]=='[' or s[i]=='{':
                st.append(s[i]);
            elif s[i]==')':
                if len(st)==0: return False
                if st[-1]!='(': return False;
                st.pop()
            elif s[i]==']':
                if len(st)==0:return False
                if st[-1]!='[':return False
                st.pop()
            else:                       #s[i]='}'
                if len(st)==0:return False
                if st[-1]!='{': return False
                st.pop()
        return len(st)==0
